### PR TITLE
[BI-XX] Middleware

### PR DESCRIPTION
**Pull request by LAlex14**
Am implementat middleware cu ajutorul VueX si pluginul createPersistedState. fiecare ruta poate avea in meta key-urile requiresNonLogged(pentru paginile care se afiseaza in cazul in care utilizatorul nu e logat ex: register, login..., iar in cazul in care e logat se redirectioneaza pe dashboard) sau requiresAuth(pentru paginile care se afiseaza in cazul in care utilizatorul e logat, ex: styleguide, dashboard(creata doar pentru exemplu), iar in cazul in care nu e logat se redirectioneaza pe login).
De asemenea se pot crea si alte middleware-uri separate ce pot fi introduse in meta.middleware: [] a fiecarei rute.

### CODE REVIEW

File: src/App.vue [hunk 1]
```diff
@@ -9,18 +9,40 @@
...
+      <button @click="logOut">Log out</button>
     </div>
     <router-view />
   </div>
-</template>
\ No newline at end of file
+</template>
+
+<script>
+// import store from '@/store/store';
```
Comment (Ceciliaa) [inline src/App.vue:32]: Remove unused code

File: src/App.vue [hunk 2]
```diff
@@ -9,18 +9,40 @@
...
+
+export default {
+  computed: {
+    getLoggedInState: function() {
```
Comment (Ceciliaa) [inline src/App.vue:36]: You do not need to use `function` word.
`getLoggedInState(){}`

File: src/App.vue [hunk 3]
```diff
@@ -9,18 +9,40 @@
...
+      return this.$store.getters.userState.loggedIn;
+    },
+  },
+  methods: {
+    logOut() {
+      this.$store.commit('setUserState');
```
Comment (Ceciliaa) [inline src/App.vue:42]: Here is an example of how we should use the getters and actions  from vuex 
image

File: src/App.vue [hunk 4]
```diff
@@ -9,18 +9,40 @@
...
+      localStorage.removeItem('vuex');
```
Comment (Ceciliaa) [inline src/App.vue:43]: It is not necessarily to remove the vuex from localStorage

File: src/components/BaseButton.vue [hunk 5]
```diff
@@ -2,18 +2,20 @@
   <button
     class="inline-flex items-center font-medium text-white rounded-md shadow-sm"
     :class="{
-      'cursor-not-allowed': disabled,
+      'cursor-not-allowed': disabled || loading,
```
Comment (Ceciliaa) [inline src/components/BaseButton.vue:5]: This code `disabled || loading` is repeating 4 times, you can write a computed method for this

File: src/router/middleware/ifLoggedInDashboardElseNext.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+export default function ifLoggedInDashboardElseNext({ next, store }) {
```
Comment (Ceciliaa) [inline src/router/middleware/ifLoggedInDashboardElseNext.js:1]: The Name of this file  should be `authMiddleware.js`

File: src/router/index.js [hunk 7]
```diff
@@ -55,4 +84,30 @@ const router = new VueRouter({
   routes
 })
 
+router.beforeEach((to, from, next) => {
```
Comment (Ceciliaa) [inline src/router/index.js:87]: This logic should be in the `authMiddleware.js` file

Comment (Ceciliaa) [reply src/router/index.js:87]: Here is an example of how  you can do that
image

File: src/views/Login.vue [hunk 8]
```diff
@@ -1,85 +1,92 @@
...
-    try {
+      try {
         this.loading = true;
-        let {data} =  await authService.login(this.model)
-        const token = get(data, 'token', '')
-        authService.setToken(token)
+        let { data } = await authService.login(this.model);
+        const token = get(data, 'token.plainTextToken', '');
+        authService.setToken(token);
+        this.$store.commit('setUserState', token);
```
Comment (Ceciliaa) [inline src/views/Login.vue:82]: Same as above, you can use mapActions and mapGetters to access the methods from vuex

File: src/router/middleware/ifLoggedInDashboardElseNext.js (hunk 6 above)
Comment (LAlex14) [reply src/router/middleware/ifLoggedInDashboardElseNext.js:1]: Asta nu mai folosește la nimic, e lăsată doar că un exemplu că se pot implementa

File: src/router/middlewarePipeline.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+function middlewarePipeline(context, middleware, index) {
```
Comment (PascuMihai) [inline src/router/middlewarePipeline.js:1]: A cleaner version would be to have all the logic in one place, let's keep the index.js just for the routes
```
let loginRoute = {
  path: '/login'
};

function redirectToLogin(next) {
  return next(loginRoute);
}

/**
 * Middleware to check if user has the correct right to access a certain page.
 * Checks are performed based on route meta fields `meta: { requiresAuth: true }`.
 * @param {object} router Vue router instance
 */
export default function authMiddleware(router) {
  router.beforeEach(async (to, from, next) => {

    const requiresAuth = to.matched.some(record => record.meta.requiresAuth);

    if (!requiresAuth) {
      return next();
    }

    const token = localStorage.getItem('token');

    if (!token) {
      return redirectToLogin(next);
    }
    return next();
  });
}
```

Comment (PascuMihai) [reply src/router/middlewarePipeline.js:1]: Then in main.js you import this function
`import authMiddleware from "@/router/authMiddleware";`
Then you pass it the router
`authMiddleware(router)`

File: src/store/store.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        }
+    },
+
+    mutations: {
+        setUserState(state, token = '') {
+            state.user.loggedIn = !!token;
+            state.user.token = token || '';
+        },
+    },
+    plugins: [createPersistedState()],
```
Comment (PascuMihai) [inline src/store/store.js:29]: FYI if we split the store into modules you can pass on objet to this function and specify the paths
```
plugins: [
    createPersistedState({
      paths: ['auth']
    })
  ],
```
the object is the options

File: src/router/middlewarePipeline.js (hunk 9 above)
Comment (PascuMihai) [reply src/router/middlewarePipeline.js:1]: The `record.meta.requiresAuth` is set in the apiConfig file
```
function isUnauthorized(status) {
  const isOnSecuredRoute = router.app.$route.matched.some(record => record.meta.requiresAuth)

  return [401, 403].includes(status) && isOnSecuredRoute
}
```

Comment (PascuMihai) [reply src/router/middlewarePipeline.js:1]:
```
export async function errorInterceptor(error) {
    const { status } = error.response;
    let errors = '';

    if (statusCodesToHandle.includes(status)) {
        errors = mapErrors(error.response.data);
    }

    error.errors = errors;

    if (isUnauthorized(status)) {
    
        await AuthService.logout();
        await app.$router.push('/login');
    
        return Promise.reject(error);
      }

    return Promise.reject(error);
}
```

Comment (PascuMihai) [reply src/router/middlewarePipeline.js:1]: Let me know if it works 👍

Comment (LAlex14) [reply src/router/middlewarePipeline.js:1]: it works :)

**Review by PascuMihai — APPROVED** [review]

File: src/components/BaseButton.vue [hunk 11]
```diff
@@ -71,6 +71,9 @@ export default {
         click: this.onClick,
       };
     },
+    disabledOrloading() {
```
Comment (Ceciliaa) [inline src/components/BaseButton.vue:74]: A better name for this method: `isDisabled`

File: src/store/store.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
+import Vue from 'vue';
+import Vuex from 'vuex';
+// import authServices from '@/api/authService';
```
Comment (Ceciliaa) [inline src/store/store.js:3]: remove this line

**Review by Ceciliaa — APPROVED** [review]

**Pull request merged**